src/librustc_typeck/collect.rs

```diff
@@ -1340,23 +1347,106 @@ fn find_existential_constraints<'a, 'tcx>(
                 .tcx
                 .typeck_tables_of(def_id)
                 .concrete_existential_types
-                .get(&self.def_id)
-                .cloned();
-            if let Some(ty) = ty {
+                .get(&self.def_id);
+            if let Some(ty::ResolvedOpaqueTy { concrete_type, substs }) = ty {
                 // FIXME(oli-obk): trace the actual span from inference to improve errors
                 let span = self.tcx.def_span(def_id);
-                if let Some((prev_span, prev_ty)) = self.found {
-                    if ty != prev_ty {
+                // used to quickly look up the position of a generic parameter
+                let mut index_map: FxHashMap<ty::ParamTy, usize> = FxHashMap::default();
+                // skip binder is ok, since we only use this to find generic parameters and their
+                // positions.
+                for (idx, subst) in substs.iter().enumerate() {
+                    if let UnpackedKind::Type(ty) = subst.unpack() {
+                        if let ty::Param(p) = ty.sty {
+                            if index_map.insert(p, idx).is_some() {
+                                // there was already an entry for `p`, meaning a generic parameter
+                                // was used twice
+                                self.tcx.sess.span_err(
+                                    span,
+                                    &format!("defining existential type use restricts existential \
+                                    type by using the generic parameter `{}` twice", p.name),
+                                );
+                                return;
+                            }
+                        } else {
+                            self.tcx.sess.delay_span_bug(
+                                span,
+                                &format!(
+                                    "non-defining exist ty use in defining scope: {:?}, {:?}",
+                                    concrete_type, substs,
+                                ),
+                            );
+                        }
+                    }
+                }
+                // compute the index within the existential type for each generic parameter used in
+                // the concrete type
+                let indices = concrete_type
+                    .subst(self.tcx, substs)
+                    .walk()
+                    .filter_map(|t| match &t.sty {
+                        ty::Param(p) => Some(*index_map.get(p).unwrap()),
+                        _ => None,
+                    }).collect();
```

Comment:
Trying to understand what this does, we have these facts:
1. `concrete_type`, in the "generics domain" of the opaque type
2. `concrete_type.subst(self.tcx, substs)`, in the "generics domain" of the definer
3. `substs` are required to be a 1:1 mapping from the opaque type to the definer
  * this implies `index_map` is the *exact inverse* of `substs`

All that means `indices` could be computed so:
```rust
                let indices = concrete_type
                    .walk()
                    .filter_map(|t| match &t.sty {
                        ty::Param(p) => Some(p.index),
                        _ => None,
                    }).collect();
```

If `indices` have to agree between definers, that means `concrete_type`s *are fully equal*.

Am I missing something? This entire PR, other than checking `substs`, looks like a noop.

I'll attempt a revert of the parts that look like a noop, and generalize to handle `const` generics (if this is even the right place to do so).

Reply:
Opened #70272.